src/ts/utils/prosemirror.ts

```diff
@@ -75,3 +78,101 @@ export const getReplaceTransactions = (tr: Transaction) =>
   tr.steps.filter(
     step => step instanceof ReplaceStep || step instanceof ReplaceAroundStep
   );
+
+interface ISuggestionFragment {
+  text: string;
+  marks: Array<Mark<any>>;
+  from: number;
+  to: number;
+}
+
+/**
+ * Generates a minimal array of replacement nodes and ranges from two pieces of text.
+ * This lets us make the smallest change to the document possible given a replacement,
+ * which gives us a better chance of preserving marks sensibly.
+ *
+ * Preserves marks that span the entirety of the text to be replaced.
+ */
+export const getReplacementFragmentsFromReplacement = (
+  tr: Transaction,
+  from: number,
+  to: number,
+  replacement: string
+): ISuggestionFragment[] => {
+  const currentText = tr.doc.textBetween(from, to);
+  const patches = jsDiff.diffChars(currentText, replacement, {
+    ignoreCase: false
+  });
+
+  const { fragments } = patches.reduce(
+    ({ fragments: currentFragments, currentPos }, patch) => {
+      // If there are no chars, ignore.
+      if (!patch.count) {
+        return {
+          fragments: currentFragments,
+          currentPos
+        };
+      }
+
+      // If this patch hasn't changed anything, ignore it and
+      // increment the count.
+      if (!patch.added && !patch.removed) {
+        return {
+          fragments: currentFragments,
+          currentPos: currentPos + patch.count
+        };
+      }
+      const $from = tr.doc.resolve(from + currentPos);
+      const $to = tr.doc.resolve(Math.min($from.pos + patch.count, tr.doc.nodeSize - 2));
+
+      // If this patch removes chars, create a fragment for
+      // the range, and leave the cursor where it is.
+      if (patch.removed) {
+        return {
+          fragments: currentFragments.concat({
+            from: $from.pos,
+            to: $to.pos,
+            text: "",
+            marks: []
+          }),
+          currentPos
+        };
+      }
+
+      const prevFragment = currentFragments[currentFragments.length - 1];
+      const isThisPatchNew =
+        !!prevFragment && (prevFragment.to || 0) < $from.pos;
+
+      let marks;
+      if (isThisPatchNew) {
+        // If this patch adds characters and the previous patch left
+        // a range intact, inherit the marks from the last character
+        // of that range.
```

Comment:
I'm struggling to decipher this comment, could you try explaining it another way? What is meant by leaving a range intact?

Reply:
Yes, this is an interesting one to explain – I've just updated that comment. Let's discuss further if that's not clear, it'd be nice to make this as comprehensible as possible!